File: typescript/tests/reasoning.test.ts
import {
  CONTRACT_VERSION,
  ContextTree,
  Format,
  isContextTreeEmpty,
  MalformedContentError,
  MalformedInputError,
  MalformedJSONError,
  MalformedReasoningFieldError,
  NodeKind,
  NodeStatus,
  ReasoningAdapter,
  ReasoningOptions,
  ReasoningTooLargeError,
  TooDeepError,
  TooManyNodesError,
  UnbalancedThinkingError,
} from "../src/reasoning/index.js";

const FIXED = new Date("2026-09-10T12:00:00Z");

function fixedClock(): Date {
  return new Date(FIXED.getTime());
}

describe("strip", () => {
  it("handles native reasoning_content", () => {
    const adapter = new ReasoningAdapter();
    const raw = {
      message: {
        role: "assistant",
        content: '{"tool": "search", "args": {"q": "x"}}',
        reasoning_content: "I need to search first.",
      },
    };
    expect(adapter.strip(raw)).toBe('{"tool": "search", "args": {"q": "x"}}');
  });

  it("strips thinking tags", () => {
    const adapter = new ReasoningAdapter();
    const raw = '<thinking>First, check the index.</thinking>\n[TOOL_CALLS] tool1: {"a":1} [/TOOL_CALLS]';
    expect(adapter.strip(raw)).toBe('\n[TOOL_CALLS] tool1: {"a":1} [/TOOL_CALLS]');
  });

  it("passes through content with no reasoning", () => {
    const adapter = new ReasoningAdapter();
    const cases = [
      '{"tool": "search", "args": {}}',
      '[{"id": "1", "name": "search", "arguments": {}}]',
      "plain text response",
      "",
    ];
    for (const raw of cases) {
      expect(adapter.strip(raw)).toBe(raw);
    }
  });

  it("handles a model-specific field", () => {
    const adapter = new ReasoningAdapter();
    adapter.registerModelField("spark", "thinking_payload");
    const raw = { content: "hello", thinking_payload: "spark reasoning" };
    expect(adapter.strip(raw)).toBe("hello");
  });
});

describe("fail-closed", () => {
  it("rejects a non-string reasoning field", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip({ reasoning_content: ["not", "a", "string"] })).toThrow(
      MalformedReasoningFieldError
    );
  });

  it("rejects a non-string content field", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip({ content: ["parts"] })).toThrow(MalformedContentError);
  });

  it("rejects unbalanced thinking tags", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip("<thinking>never closed")).toThrow(UnbalancedThinkingError);
  });

  it("rejects a dangling close tag", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip("no open tag [/THINK]")).toThrow(UnbalancedThinkingError);
  });

  it("rejects malformed JSON that looks structured", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip('{"message": {"content": "x"}')).toThrow(MalformedJSONError);
  });

  it("rejects unsupported types", () => {
    const adapter = new ReasoningAdapter();
    expect(() => adapter.strip(42)).toThrow(MalformedInputError);
  });
});

describe("bounds", () => {
  it("rejects reasoning over the byte cap", () => {
    const adapter = new ReasoningAdapter({ max_reasoning_bytes: 16 });
    expect(() =>
      adapter.extract({ content: "", reasoning_content: "x".repeat(64) }, "deepseek", "test")
    ).toThrow(ReasoningTooLargeError);
  });

  it("rejects too many nodes", () => {
    const adapter = new ReasoningAdapter({ max_nodes: 2, clock: fixedClock });
    expect(() => adapter.extract("<thinking>p1\n\np2\n\np3\n\np4</thinking>", "m", "b")).toThrow(
      TooManyNodesError
    );
  });

  it("rejects a chain that is too deep", () => {
    const adapter = new ReasoningAdapter({ max_depth: 2, clock: fixedClock });
    expect(() => adapter.extract("<thinking>a\n\nb\n\nc</thinking>", "m", "b")).toThrow(
      TooDeepError
    );
  });
});

describe("extract", () => {
  it("builds a versioned context tree", () => {
    const adapter = new ReasoningAdapter({ clock: fixedClock });
    const raw =
      "<thinking>" +
      "goal: verify the payment gateway\n\n" +
      "observation: gateway returned 200 [E1]\n\n" +
      "decision: call refund tool\n\n" +
      'tool-call: refund_tool: {"id": "txn1"}\n\n' +
      "tool-result: refund_tool: ok [SUPERSEDED]\n\n" +
      "conclusion: refund issued\n\n" +
      "blocker: regulatory approval pending" +
      "</thinking>";

    const tree = adapter.extract(raw, "deepseek-reasoner", "test-backend");

    expect(tree.version).toBe(CONTRACT_VERSION);
    expect(tree.format).toBe(Format.THINKING);
    expect(tree.model).toBe("deepseek-reasoner");
    expect(tree.backend).toBe("test-backend");
    expect(tree.root_id).not.toBe("");

    expect(tree.nodes).toHaveLength(7);
    expect(tree.nodes.map((n) => n.kind)).toEqual([
      NodeKind.GOAL,
      NodeKind.OBSERVATION,
      NodeKind.DECISION,
      NodeKind.TOOL_CALL,
      NodeKind.TOOL_RESULT,
      NodeKind.CONCLUSION,
      NodeKind.BLOCKER,
    ]);

    // Chain: root has empty parent, every other node's parent is its predecessor.
    expect(tree.nodes[0].parent_id).toBe("");
    for (let i = 1; i < tree.nodes.length; i++) {
      expect(tree.nodes[i].parent_id).toBe(tree.nodes[i - 1].id);
    }

    expect(tree.nodes[0].status).toBe(NodeStatus.COMPLETE);
    expect(tree.nodes[4].status).toBe(NodeStatus.SUPERSEDED);
    expect(tree.nodes[6].status).toBe(NodeStatus.BLOCKED);

    expect(tree.nodes[1].evidence_refs).toEqual(["E1"]);
    expect(tree.nodes[3].tool_call_refs).toEqual(["refund_tool"]);

    for (const node of tree.nodes) {
      expect(node.created_at.toISOString()).toBe(FIXED.toISOString());
      expect(node.updated_at.toISOString()).toBe(FIXED.toISOString());
    }
  });

  it("returns an empty tree when there is no reasoning", () => {
    const adapter = new ReasoningAdapter({ clock: fixedClock });
    const tree = adapter.extract({ content: "hello" }, "m", "b");
    expect(tree.nodes).toHaveLength(0);
    expect(tree.root_id).toBe("");
    expect(tree.format).toBe(Format.NONE);
  });

  it("produces deterministic node ids", () => {
    const adapter = new ReasoningAdapter({ clock: fixedClock });
    const raw = "<thinking>observation: x\n\ndecision: y</thinking>";
    const t1 = adapter.extract(raw, "m", "b");
    const t2 = adapter.extract(raw, "m", "b");
    expect(t1.nodes.map((n) => n.id)).toEqual(t2.nodes.map((n) => n.id));
    expect(t1.root_id).toBe(t2.root_id);
  });

  it("bounds summaries and never stores raw reasoning", () => {
    const adapter = new ReasoningAdapter({ max_summary_chars: 10, clock: fixedClock });
    const raw = "<thinking>observation: " + "abcdefg".repeat(20) + "</thinking>";
    const tree = adapter.extract(raw, "m", "b");
    expect(tree.nodes).toHaveLength(1);
    expect(Array.from(tree.nodes[0].summary).length).toBeLessThanOrEqual(10);
    expect(tree.nodes[0].summary).not.toContain("abcdefg".repeat(20));
  });

  it("truncates on code-point boundaries", () => {
    const adapter = new ReasoningAdapter({ max_summary_chars: 3, clock: fixedClock });
    const raw = "<thinking>observation: héllo wörld 🚀 end</thinking>";
    const tree = adapter.extract(raw, "m", "b");
    expect(tree.nodes[0].summary).toBe("hél");
  });
});

describe("parse", () => {
  it("returns clean content and the tree", () => {
    const adapter = new ReasoningAdapter({ clock: fixedClock });
    const raw = "<thinking>observation: x</thinking>\n" + '{"tool": "search", "args": {"q": "1"}}';
    const { clean, tree } = adapter.parse(raw, "m", "b");
    expect(clean).toBe('\n{"tool": "search", "args": {"q": "1"}}');
    expect(tree.nodes.length).toBeGreaterThan(0);
  });
});

describe("contract", () => {
  it("exposes the closed vocabularies", () => {
    expect(Object.values(NodeKind)).toEqual([
      "goal",
      "observation",
      "decision",
      "tool-call",
      "tool-result",
      "conclusion",
      "blocker",
    ]);
    expect(Object.values(NodeStatus)).toEqual(["active", "complete", "blocked", "superseded"]);
    expect(Object.values(Format)).toEqual(["native", "thinking", "model_specific", "none"]);
  });

  it("serialized tree never contains whole raw reasoning", () => {
    const adapter = new ReasoningAdapter({ clock: fixedClock });
    const raw = "<thinking>observation: " + "z".repeat(1000) + "</thinking>";
    const tree = adapter.extract(raw, "m", "b");
    const serialized = JSON.stringify(tree);
    expect(serialized).not.toContain("z".repeat(1000));
    expect(Array.from(tree.nodes[0].summary).length).toBeLessThanOrEqual(512);
  });
});

describe("isContextTreeEmpty", () => {
  it("reports empty trees", () => {
    expect(isContextTreeEmpty(null)).toBe(true);
    expect(isContextTreeEmpty({ nodes: [] } as unknown as ContextTree)).toBe(true);
    const node: import("../src/reasoning/types.js").Node = {
      id: "a",
      parent_id: "",
      kind: NodeKind.GOAL,
      status: NodeStatus.COMPLETE,
      summary: "x",
      evidence_refs: [],
      tool_call_refs: [],
      created_at: FIXED,
      updated_at: FIXED,
    };
    const tree: ContextTree = {
      version: CONTRACT_VERSION,
      format: Format.NONE,
      model: "m",
      backend: "b",
      created_at: FIXED,
      root_id: "a",
      nodes: [node],
    };
    expect(isContextTreeEmpty(tree)).toBe(false);
  });
});

File: python/tests/reasoning_test.py
"""Tests for the reasoning adapter (mirrors go/reasoning/adapter_test.go)."""

from datetime import datetime, timezone

import pytest

from pedro_agentware.reasoning import (
    CONTRACT_VERSION,
    ContextTree,
    Format,
    MalformedContentError,
    MalformedInputError,
    MalformedJSONError,
    MalformedReasoningFieldError,
    Node,
    NodeKind,
    NodeStatus,
    ReasoningAdapter,
    ReasoningOptions,
    ReasoningTooLargeError,
    TooDeepError,
    TooManyNodesError,
    UnbalancedThinkingError,
)

FIXED = datetime(2026, 9, 10, 12, 0, 0, tzinfo=timezone.utc)


def fixed_clock() -> datetime:
    return FIXED


class TestStrip:
    def test_native_reasoning_content(self):
        adapter = ReasoningAdapter()
        raw = {
            "message": {
                "role": "assistant",
                "content": '{"tool": "search", "args": {"q": "x"}}',
                "reasoning_content": "I need to search first.",
            }
        }
        assert adapter.strip(raw) == '{"tool": "search", "args": {"q": "x"}}'

    def test_thinking_tags(self):
        adapter = ReasoningAdapter()
        raw = '<thinking>First, check the index.</thinking>\n[TOOL_CALLS] tool1: {"a": 1} [/TOOL_CALLS]'
        assert adapter.strip(raw) == '\n[TOOL_CALLS] tool1: {"a": 1} [/TOOL_CALLS]'

    def test_no_reasoning_passthrough(self):
        adapter = ReasoningAdapter()
        cases = [
            '{"tool": "search", "args": {}}',
            '[{"id": "1", "name": "search", "arguments": {}}]',
            "plain text response",
            "",
        ]
        for raw in cases:
            assert adapter.strip(raw) == raw

    def test_model_specific_field(self):
        adapter = ReasoningAdapter()
        adapter.register_model_field("spark", "thinking_payload")
        raw = {"content": "hello", "thinking_payload": "spark reasoning"}
        assert adapter.strip(raw) == "hello"

    def test_json_string_input_is_opaque(self):
        adapter = ReasoningAdapter()
        raw = '[{"id": "1", "name": "search", "arguments": {"q": "x"}}]'
        assert adapter.strip(raw) == raw


class TestFailClosed:
    def test_non_string_reasoning_field(self):
        adapter = ReasoningAdapter()
        with pytest.raises(MalformedReasoningFieldError):
            adapter.strip({"reasoning_content": ["not", "a", "string"]})

    def test_non_string_content(self):
        adapter = ReasoningAdapter()
        with pytest.raises(MalformedContentError):
            adapter.strip({"content": ["parts"]})

    def test_unbalanced_thinking_tag(self):
        adapter = ReasoningAdapter()
        with pytest.raises(UnbalancedThinkingError):
            adapter.strip("<thinking>never closed")

    def test_dangling_close_tag(self):
        adapter = ReasoningAdapter()
        with pytest.raises(UnbalancedThinkingError):
            adapter.strip("no open tag [/THINK]")

    def test_malformed_json(self):
        adapter = ReasoningAdapter()
        with pytest.raises(MalformedJSONError):
            adapter.strip('{"message": {"content": "x"}')

    def test_unsupported_type(self):
        adapter = ReasoningAdapter()
        with pytest.raises(MalformedInputError):
            adapter.strip(42)


class TestBounds:
    def test_reasoning_too_large(self):
        adapter = ReasoningAdapter(ReasoningOptions(max_reasoning_bytes=16))
        raw = {"content": "", "reasoning_content": "x" * 64}
        with pytest.raises(ReasoningTooLargeError):
            adapter.extract(raw, "deepseek", "test")

    def test_too_many_nodes(self):
        adapter = ReasoningAdapter(ReasoningOptions(max_nodes=2, clock=fixed_clock))
        with pytest.raises(TooManyNodesError):
            adapter.extract("<thinking>p1\n\np2\n\np3\n\np4</thinking>", "m", "b")

    def test_too_deep(self):
        adapter = ReasoningAdapter(ReasoningOptions(max_depth=2, clock=fixed_clock))
        with pytest.raises(TooDeepError):
            adapter.extract("<thinking>a\n\nb\n\nc</thinking>", "m", "b")


class TestExtract:
    def test_context_tree(self):
        adapter = ReasoningAdapter(ReasoningOptions(clock=fixed_clock))
        raw = (
            "<thinking>"
            "goal: verify the payment gateway\n\n"
            "observation: gateway returned 200 [E1]\n\n"
            "decision: call refund tool\n\n"
            'tool-call: refund_tool: {"id": "txn1"}\n\n'
            "tool-result: refund_tool: ok [SUPERSEDED]\n\n"
            "conclusion: refund issued\n\n"
            "blocker: regulatory approval pending"
            "</thinking>"
        )

        tree = adapter.extract(raw, "deepseek-reasoner", "test-backend")

        assert tree.version == CONTRACT_VERSION
        assert tree.format == Format.THINKING
        assert tree.model == "deepseek-reasoner"
        assert tree.backend == "test-backend"
        assert tree.root_id != ""

        assert len(tree.nodes) == 7
        expected_kinds = [
            NodeKind.GOAL,
            NodeKind.OBSERVATION,
            NodeKind.DECISION,
            NodeKind.TOOL_CALL,
            NodeKind.TOOL_RESULT,
            NodeKind.CONCLUSION,
            NodeKind.BLOCKER,
        ]
        assert [n.kind for n in tree.nodes] == expected_kinds

        # Chain: root has empty parent, every other node's parent is its predecessor.
        assert tree.nodes[0].parent_id == ""
        for prev, node in zip(tree.nodes, tree.nodes[1:]):
            assert node.parent_id == prev.id

        assert tree.nodes[0].status == NodeStatus.COMPLETE
        assert tree.nodes[4].status == NodeStatus.SUPERSEDED
        assert tree.nodes[6].status == NodeStatus.BLOCKED

        assert tree.nodes[1].evidence_refs == ["E1"]
        assert tree.nodes[3].tool_call_refs == ["refund_tool"]

        assert all(n.created_at == FIXED for n in tree.nodes)
        assert all(n.updated_at == FIXED for n in tree.nodes)

    def test_empty_reasoning(self):
        adapter = ReasoningAdapter(ReasoningOptions(clock=fixed_clock))
        tree = adapter.extract({"content": "hello"}, "m", "b")
        assert tree.empty()
        assert tree.root_id == ""
        assert tree.format == Format.NONE

    def test_deterministic_node_ids(self):
        adapter = ReasoningAdapter(ReasoningOptions(clock=fixed_clock))
        raw = "<thinking>observation: x\n\ndecision: y</thinking>"
        t1 = adapter.extract(raw, "m", "b")
        t2 = adapter.extract(raw, "m", "b")
        assert [n.id for n in t1.nodes] == [n.id for n in t2.nodes]
        assert t1.root_id == t2.root_id

    def test_summary_bounded(self):
        adapter = ReasoningAdapter(ReasoningOptions(max_summary_chars=10, clock=fixed_clock))
        raw = "<thinking>observation: " + "abcdefg" * 20 + "</thinking>"
        tree = adapter.extract(raw, "m", "b")
        assert len(tree.nodes) == 1
        assert len(tree.nodes[0].summary) <= 10
        assert "abcdefg" * 20 not in tree.nodes[0].summary

    def test_unicode_summary_truncation(self):
        adapter = ReasoningAdapter(ReasoningOptions(max_summary_chars=3, clock=fixed_clock))
        raw = "<thinking>observation: héllo wörld 🚀 end</thinking>"
        tree = adapter.extract(raw, "m", "b")
        assert tree.nodes[0].summary == "hél"


class TestParse:
    def test_returns_clean_and_tree(self):
        adapter = ReasoningAdapter(ReasoningOptions(clock=fixed_clock))
        raw = "<thinking>observation: x</thinking>\n" + '{"tool": "search", "args": {"q": "1"}}'
        clean, tree = adapter.parse(raw, "m", "b")
        assert clean == '\n{"tool": "search", "args": {"q": "1"}}'
        assert not tree.empty()


class TestContract:
    def test_vocabulary(self):
        assert [k.value for k in NodeKind] == [
            "goal",
            "observation",
            "decision",
            "tool-call",
            "tool-result",
            "conclusion",
            "blocker",
        ]
        assert [s.value for s in NodeStatus] == [
            "active",
            "complete",
            "blocked",
            "superseded",
        ]
        assert [f.value for f in Format] == [
            "native",
            "thinking",
            "model_specific",
            "none",
        ]

    def test_to_dict_never_contains_raw_reasoning(self):
        adapter = ReasoningAdapter(ReasoningOptions(clock=fixed_clock))
        # Raw reasoning longer than the summary bound must never appear whole.
        raw = "<thinking>observation: " + "z" * 1000 + "</thinking>"
        tree = adapter.extract(raw, "m", "b")
        serialized = tree.to_dict()
        assert "z" * 1000 not in str(serialized)
        assert len(tree.nodes[0].summary) <= 512


class TestContextTreeEmpty:
    def test_empty(self):
        assert ContextTree().empty()
        assert ContextTree(nodes=[]).empty()
        assert not ContextTree(
            nodes=[_node()],
        ).empty()


def _node() -> Node:
    return Node(
        id="abc",
        parent_id="",
        kind=NodeKind.GOAL,
        status=NodeStatus.COMPLETE,
        summary="goal",
        created_at=FIXED,
        updated_at=FIXED,
    )
